File: Binary Search/search-in-rotated-sorted-array.py
# 32ms 98.25%
# 13.2MB 55.05%
class Solution:

    def search(self, nums, target) -> int:
        if len(nums) == 0:
            return -1
        index = self.findIndex(nums, 0, len(nums) - 1)
        if index is -1:
            return self.searchInSubnums(nums, target, 0 , len(nums)-1, 0)
        if target <= nums[-1]:
            return self.searchInSubnums(nums[index:], target, 0 , len(nums[index:])-1, len(nums[:index]))
        else:
            return self.searchInSubnums(nums[:index], target, 0 , len(nums[:index])-1, 0)

    def findIndex(self, nums, left, right):

        if left + 1 >= right:
            if nums[left] > nums[right]:
                return right
            else:
                return -1

        k = int((left + right) / 2)
        if nums[k] > nums[right]:
            return self.findIndex(nums, k, right)
        elif nums[k] < nums[left]:
            return self.findIndex(nums, left, k)
        return -1

    def searchInSubnums(self, nums, target, left, right, re_len):

        k = int((left+right)/2)
        if nums[k] == target:
            return k + re_len
        if left >= right and nums[k] != target:
            return -1
        if nums[k] < target:
            return self.searchInSubnums(nums, target, k+1, right, re_len)
        elif nums[k] > target:
            return self.searchInSubnums(nums, target, left, k-1, re_len)

nums = [1,3,5]
a = Solution()
print(a.search(nums,1))

File: Array/search-in-rotated-sorted-array.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int search(vector<int>& nums, int target) {
        if (nums.size() == 1) return nums[0] == target ? 0 : -1;

        // search the split
        int left = 0, right = nums.size()-1;
        while (left < right){
            int mid = left + (right - left) / 2;
            if (nums[mid] > nums[right]) left = mid + 1;
            else right = mid;
        }
        // search the target
        int min_index, max_index;
        if (left == 0) min_index = 0, max_index = nums.size()-1;
        else min_index = left, max_index = left - 1;
        if (target > nums[max_index] || target < nums[min_index]) return -1;

        if (target > nums[nums.size()-1]){
            left = 0;
            right = right - 1;
        } else{
            left = left;
            right = nums.size()-1;
        }
        while (left <= right){
            int mid = left + (right - left) / 2;
            if (nums[mid] == target) return mid;
            if (nums[mid] < target) left = mid + 1;
            else right = mid - 1;
        }
        return -1;
    }
};

int main() {
    Solution sol;
    vector<int> nums = {1, 3};
    int target = 2;
    int ans = sol.search(nums, target);
    cout << ans << endl;
    return 0;
}